src/plugins/data/server/search/session/session_service.ts

```diff
@@ -0,0 +1,198 @@
+    name: string,
+    url: string,
+    userId: string,
+    expires: Date = new Date(Date.now() + DEFAULT_EXPIRATION),
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    // Get the mapping of request hash/search ID for this session
+    const searchMap = this.sessionSearchMap.get(sessionId) ?? new Map<string, string>();
+
+    const idMapping = Object.fromEntries(searchMap.entries());
+    const attributes = { name, url, userId, expires: expires.toISOString(), idMapping };
+    const session = savedObjectsClient.create<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      attributes,
+      { id: sessionId }
+    );
+
+    // Clear out the entries for this session ID so they don't get saved next time
+    this.sessionSearchMap.delete(sessionId);
+
+    return session;
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public get = (sessionId: string, { savedObjectsClient }: BackgroundSessionDependencies) => {
+    return savedObjectsClient.get<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      sessionId
+    );
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public find = (
+    options: SearchSessionFindOptions,
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    return savedObjectsClient.find<BackgroundSessionSavedObjectAttributes>({
+      ...options,
+      type: BACKGROUND_SESSION_TYPE,
+    });
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public update = (
+    sessionId: string,
+    attributes: Partial<BackgroundSessionSavedObjectAttributes>,
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    return savedObjectsClient.update<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      sessionId,
+      attributes
+    );
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public delete = (sessionId: string, { savedObjectsClient }: BackgroundSessionDependencies) => {
+    return savedObjectsClient.delete(BACKGROUND_SESSION_TYPE, sessionId);
+  };
+
+  /**
+   * Tracks the given search request/search ID in the saved session (if it exists). Otherwise, just
+   * store it in memory until a saved session exists.
+   * @internal
+   */
+  public trackId = async (
+    searchRequest: IKibanaSearchRequest,
+    searchId: string,
+    { sessionId, isStored }: ISearchOptions,
+    deps: BackgroundSessionDependencies
+  ) => {
+    if (!sessionId || !searchId) return;
+    const requestHash = createRequestHash(searchRequest.params);
+
+    // If there is already a saved object for this session, update it to include this request/ID.
+    // Otherwise, just update the in-memory mapping for this session for when the session is saved.
+    if (isStored) {
+      const attributes = { idMapping: { [requestHash]: searchId } };
+      await this.update(sessionId, attributes, deps);
```

Comment:
Just double checking: 
Would update here completely override attributes or would it add new attributes.idMapping[requestHash] entry? 
I assume the later is expected, correct?

Reply:
It's a partial update, so it only adds additional mappings. I've verified this is how it works.